stack/clone.go

```diff
@@ -101,3 +127,104 @@ func copyFile(destfile, srcfile string) error {
 	_, err = io.Copy(dest, src)
 	return err
 }
+
+func updateStackID(stackdir string) error {
+	logger := log.With().
+		Str("action", "stack.updateStackID()").
+		Str("stack", stackdir).
+		Logger()
+
+	logger.Trace().Msg("parsing stack")
+
+	parser, err := hcl.NewTerramateParser(stackdir, stackdir)
+	if err != nil {
+		return err
+	}
+
+	if err := parser.AddDir(stackdir); err != nil {
+		return err
+	}
+
+	if err := parser.Parse(); err != nil {
+		return err
+	}
+
+	logger.Trace().Msg("finding file containing stack definition")
+
+	stackFilePath, body := getStackBody(parser)
+	if body == nil {
+		return errors.E("updating stack ID: stack block not found")
+	}
+
+	// WHY oh WHY do you ask ? Parsing HCL always delivers an AST that
+	// has no comments on it, so building a new HCL file from the parsed
+	// AST will lose all comments from the original code.
+	//
+	// - https://raw.githubusercontent.com/katcipis/memes/master/satan.jpg
+
+	logger.Trace().Msg("reading cloned stack file")
+
+	stackContents, err := os.ReadFile(stackFilePath)
+	if err != nil {
+		return errors.E(err, "reading cloned stack definition file")
+	}
+
+	logger.Trace().Msg("lexing cloned stack file")
+
+	roTokens, err := lex.Config(stackContents, stackFilePath)
+	if err != nil {
+		return errors.E(err, "cloned stack is invalid HCL")
+	}
+
+	logger.Trace().Msg("finding stack block on tokens")
+
+	tokens := lex.WriterTokens(roTokens)
+	blockStart, ok := lex.FindTokenSequence(tokens, lex.TokenIdent("stack"), lex.TokenOBrace())
+	if !ok {
+		return errors.E(err, "cloned stack doesn't have stack block")
+	}
+
+	logger.Trace().Msg("finding id attribute")
+
+	// We can assume at this point that the stack has an ID, since previous parsing checked that.
+	// This is not generally safe, if we allow the stack block to have attributes that
+	// have objects as values and those objects can have an "id" field this will fail.
+	// If we allow multiple stack blocks, this will also fail.
+	// For now we are assuming stack blocks are constrained, for something safer we need
+	// a more proper parser instead of YOLO lexing.
+
+	idAttributeOffset, ok := lex.FindTokenSequence(tokens[blockStart:], lex.TokenIdent("id"), lex.TokenEqual())
+	if !ok {
+		return errors.E(err, "cloned stack doesn't have stack ID")
+	}
+
+	logger.Trace().Msg("updating id attribute")
+
+	// Here we assume that stack IDs are also on the form:
+	// id = "id"
+	// Id's are very constrained and are always strings, so we expect
+	// the tokens: TokenOQuote + TokenQuotedLit + TokenCQuote
```

Comment:
theoretically... we documented IDs as being strings, then there's a lot of valid cases out there...

Expressions can be parenthesized without any semantic changes, so the examples below are valid ids, no?

```
id = ("id")
id = (("id"))
```

pure interpolations are automatically handled by hcl expr.Value(nil) evaluator.. so the following is also valid?

```
id = "${"part 1"} part 2"
```

Reply:
hmm if after eval they satisfy the ID constraints I think they would be, so maybe there are more cases to handle here then.

Reply:
maybe we can add support for other forms of defining the stack ID on follow up PRs, the most common cases today is no ID or the UUID we generate (plain string). so we can have something usable soon and improve (the command itself will be experimental too).